Active Anti-Entropy

jtuple opened the pull request:
Had trouble merging and rebasing, so I ended up squashing into a single commit. So, here's AAE in all it's a glory, as a single super commit. Given how much this has been rewritten/reorganized, the old commit history doesn't help much anyway.

I'm extending/polishing my riak_tests some more, but will have thoses pushed soon and linked to this pull-request.

Hopefully all documentation and included -specs make this straightforward to review, despite being a massive pull-request. Also, everything from my AAE screencast that I previously distributed is still valid:
http://coffee.jtuple.com/video/AAE.html

I'm assuming @rzezeski and @Vagabond will be the two reviewing this, but others can jump on as well. However, we should have at least 2 reviewers on this one.

AAE is currently disabled by default, you'll need to enable it in the riak_kv section in app.config.

``` erlang
{anti_entropy, {on, []}},
%% or
{anti_entropy, {on, [debug]}},
%% or
{anti_entropy, {manual, []}},
%% or
{anti_entropy, {manual, [debug]}},
```

Also, additional settings you can set/change as desired (times are in milliseconds):

``` erlang
            {anti_entropy_timeout, 20000},
            {anti_entropy_tick, 1000},
            {anti_entropy_concurrency, 2},
            {anti_entropy_expire, 10000},
```

(cabal: basho/cabal#1)

---

Active anti-entropy is a process through which multiple replicas periodically sync with each other and automatically repair any keys that are missing or have divergent data. For example, a user could delete all data from a node and Riak would automatically regenerate the data from the other replicas.

This implementation uses hash trees / Merkle trees to perform lightweight key exchange, with work proportional to the number of divergent keys rather than the size of the overall keyspace.

This implementation meets several design goals:
1. The underlying hash trees are on-disk structures, and can scale to billions+ keys. This is in contrast to in-memory trees that require significant RAM to support massive keyspaces.
2. The underlying hash trees are persistent. Riak nodes can be restarted without fear of hash tree data being lost and needing to be rebuilt.
3. As new data is written to Riak, the hash trees associated with the various partitions are kept up-to-date. Each write in Riak triggers an asynchronous write to one or more hash trees. Combined with 2, this enables trees to be built once through a scan over existing data, and then maintained in real-time. In reality, trees are expired over time and rebuilt to ensure the hash tree and backend data stay in sync, and also to identify bit rot / disk failure.
4. The entire implementation is designed to be non-blocking. For example, a snapshot of a hash tree is generated before performing an exchange with other replicas, therefore allowing concurrent inserts of new key/hash pairs as new writes occur.

The current implementation triggers read repair for each key difference identified through the hash exchange. This is a reasonable approach as read repair is stable, production-tested mechanism in Riak. However, the read repair approach leads to slower replica repair in cases where there are a large number of key differences. This is an area for future improvement.

jtuple committed:
Add active anti-entropy to Riak

Active anti-entropy is a process through which multiple replicas
periodically sync with each other and automatically repair any keys
that are missing or have divergent data. For example, a user could
delete all data from a node and Riak would automatically regenerate
the data from the other replicas.

This implementation uses hash trees / Merkle trees to perform
lightweight key exchange, with work proportional to the number of
divergent keys rather than the size of the overall keyspace.

This implementation meets several design goals:

1. The underlying hash trees are on-disk structures, and can scale to
   billions+ keys. This is in contrast to in-memory trees that require
   significant RAM to support massive keyspaces.

2. The underlying hash trees are persistent. Riak nodes can be
   restarted without fear of hash tree data being lost and needing to
   be rebuilt.

3. As new data is written to Riak, the hash trees associated with the
   various partitions are kept up-to-date. Each write in Riak triggers
   an asynchronous write to one or more hash trees. Combined with #2,
   this enables trees to be built once through a scan over existing
   data, and then maintained in real-time. In reality, trees are expired
   over time and rebuilt to ensure the hash tree and backend data stay
   in sync, and also to identify bit rot / disk failure.

4. The entire implementation is designed to be non-blocking. For
   example, a snapshot of a hash tree is generated before performing
   an exchange with other replicas, therefore allowing concurrent
   inserts of new key/hash pairs as new writes occur.

The current implementation triggers read repair for each key difference
identified through the hash exchange. This is a reasonable approach as
read repair is stable, production-tested mechanism in Riak. However,
the read repair approach leads to slower replica repair in cases where
there are a large number of key differences. This is an area for future
improvement.

rzezeski commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -0,0 +1,547 @@
+    LocalIndexN = riak_kv_util:responsible_preflists(Index, Ring),
+    Sibs = riak_kv_util:preflist_siblings(Index),
+    lists:flatmap(
+      fun(RemoteIdx) when RemoteIdx == Index ->
```
Why not just have `preflist_siblings` not include the index itself?  It removes this filter case, which confused me the first time I reviewed this code.  It seems to me an index shouldn't be a sibling to itself.

rzezeski commented on src/riak_kv_util.erl:
```diff
@@ -157,6 +165,92 @@ mapred_system() ->
+    [{Idx, N} || Idx <- lists:reverse(Pred)].
+
+-spec determine_max_n(riak_core_ring()) -> pos_integer().
+determine_max_n(Ring) ->
```
How about calling `determine_all_n` and using `lists:max`?

rzezeski commented on src/riak_kv_sup.erl:
```diff
@@ -118,6 +123,7 @@ init([]) ->
         KeysFsmSup,
         IndexFsmSup,
         LegacyKeysFsmSup,
+        ?IF(EntropyEnabled, EntropyManager, []),
```
So if a user decides they want to enable entropy now they need to restart Riak for that to take effect.  Should we add a function to entropy mgr to allow dynamic adding of child?

rzezeski commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -0,0 +1,547 @@
+    get_lock(Type, self()).
+
+%% @doc Acquire an exchange concurrency lock if available, and associate
+%%      the lock with the calling process.
```
s/with the calling process/with the `Pid'/

rzezeski commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -0,0 +1,547 @@
+            [lager:trace_console([{module, Mod}]) || Mod <- Modules];
+        false ->
+            [begin
+                 {ok, Trace} = lager:trace_console([{module, Mod}]),
```
I guess these have to be started first or else `lager:stop_trace` will error?

rzezeski commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -0,0 +1,547 @@
+    case app_helper:get_env(riak_kv, anti_entropy, {off, []}) of
+        {on, Opts} ->
+            {true, Opts};
+        {manual, Opts} ->
```
I think allowing `manual` as an entry for the first element in the tuple is a bit confusing.  Entropy mgr has the notion of automatic or manual.  That makes sense.  Either mode requires that anti entropy be "on."  To me it makes more sense if the first atom is `on | off` and manual mode is invoked by include the option `manual` or `manual_mode`.  As it stands now the user could invoke manual mode in two ways `{on, [manual]}` or `{manual, []}`.  Why have two different ways of putting the system in the same state?

rzezeski commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -0,0 +1,547 @@
+init([]) ->
+    schedule_tick(),
+    {_, Opts} = settings(),
+    Mode = case lists:member(manual, Opts) of
```
Nitpicky...why not use proplists for the options to make the intention more clear.

rzezeski commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -0,0 +1,547 @@
+    case lists:member(E, Exchanges) of
+        true ->
+            trigger_exchanges([E], State);
+        false ->
```
I feel like we shouldn't be quiet about invalid exchanges in
`trigger_exchange`.  I assume that this can only happen when a user is
triggering exchange manually.  In that case it would be helpful to log
something to let the user know it's invalid.  Also, it would be a good
safeguard just in case the situation ever arises where invalid
exchanges are trigger during automatic mode.

rzezeski commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -0,0 +1,547 @@
+    State3.
+
+
+-spec do_get_lock(any(),pid(),state()) -> {'max_concurrency',state()}.
```
s/'max_concurrency'/'ok | max_concurrency/

rzezeski commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -0,0 +1,547 @@
+    MissingIdx = [Idx || Idx <- Indices,
+                         not dict:is_key(Idx, Existing)],
+    L = [{Idx, Pid} || Idx <- MissingIdx,
+                       {ok,Pid} <- [riak_kv_vnode:hashtree_pid(Idx)],
```
If a vnode is backed up then `riak_kv_vnode:hashtree_pid` could take a
long time to respond.  This would block `reload_hashtrees` which would
also block the tick.  Do we want to block the entropy manager because
of one backed-up vnode?  Why not use a small timeout and if the vnode
doesn't return in time then it will be tried again on the next tick.

rzezeski commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -0,0 +1,547 @@
+                  end, Indices).
+
+-spec add_index_exchanges(index(), state()) -> state().
+add_index_exchanges(_Index, State) when State#state.mode == manual ->
```
Why have `add_index_exchanges` and `trigger_exchange`?  I feel like
`add_index_exchange` makes more sense because exchanges are being
added to the exchange queue (so perhaps `queue_exchange` would make
most sense).  The word "trigger" makes me thing it's happening
immediately.

rzezeski commented:
I think we are going to need to give AAE more sensible defaults for when it is turned on.  I ran read/update benchmark against a local devrel and turning AAE on with the defaults caused throughput to drop by more than half.  95th gets shot up by a factor of 3 and updates by a factor of 6.  There was also more variance with AAE on than off.  I'm going to do some runs while playing with the various knobs and see what I find.

EDIT: This was against a healthy cluster.  No attempt was made to actually cause divergence.  I simply wanted to see how the overhead introduced by AAE would affect a get/update load.

rzezeski commented:
If a node is down I see constant chatter about timeout on exchange.

```
012-11-06 18:18:03.876 [info] <0.5232.6>@riak_kv_exchange_fsm:do_timeout:255 Timeout during exchange between (local) {1004782375664995756265033322492444576013453623296,'dev1@127.0.0.1'} and (remote) {1050454301831586472458898473514828420377701515264,'dev3@127.0.0.1'}, (preflist) {1004782375664995756265033322492444576013453623296,3}
2012-11-06 18:18:03.876 [info] <0.5238.6>@riak_kv_exchange_fsm:do_timeout:255 Timeout during exchange between (local) {1187470080331358621040493926581979953470445191168,'dev1@127.0.0.1'} and (remote) {1233142006497949337234359077604363797834693083136,'dev3@127.0.0.1'}, (preflist) {1187470080331358621040493926581979953470445191168,3}
```

rzezeski commented:
> Repaired 17 keys through active anti-entropy

I'm watching these fly by with no idea what is being repaired.  It would be nice to know the nodes/index/{index,n} involved

rzezeski commented on src/riak_kv_index_hashtree.erl:
```diff
@@ -0,0 +1,592 @@
+%%      that is managed by the provided index_hashtree.
+-spec exchange_bucket(index_n(), integer(), integer(), pid()) -> orddict().
+exchange_bucket(Id, Level, Bucket, Tree) ->
+    put(calling, ?LINE),
```
What is this proc dict entry for?  Something to do with logging I take it?

rzezeski commented on src/riak_kv_index_hashtree.erl:
```diff
@@ -0,0 +1,592 @@
+%%%===================================================================
+
+init([Index, IndexNs]) ->
+    Root = "data/anti",
```
This needs to be configurable like our other data dirs.

rzezeski commented on src/riak_kv_vnode.erl:
```diff
@@ -258,7 +332,8 @@ init([Index]) ->
                     %% Create worker pool initialization tuple
                     FoldWorkerPool = {pool, riak_kv_worker, WorkerPoolSize, []},
-                    {ok, State, [FoldWorkerPool]};
+                    State2 = maybe_create_hashtrees(State),
```
The hashtree is linked to the vnode.  If the hashtree dies then the
vnode dies.  This seems wrong.  I feel like the HTs should be under
their own supervisor.  The vnode would tell the supervisor to start a
new child and then restart it as needed.

rzezeski commented on src/riak_kv_index_hashtree.erl:
```diff
@@ -0,0 +1,592 @@
+    State2 = init_trees(IndexNs, State),
+    {ok, State2}.
+
+handle_call({new_tree, Id}, _From, State) ->
```
This looks like an unused code path.

rzezeski commented on src/riak_kv_index_hashtree.erl:
```diff
@@ -0,0 +1,592 @@
+
+-spec do_get_lock(any(), pid(), state()) -> {not_built | ok | already_locked, state()}.
+do_get_lock(_, _, State) when State#state.built /= true ->
+    lager:debug("Not built: ~p :: ~p", [State#state.index, State#state.built]),
```
The `state.built` in this msg is redundant since it starts with `Not built...`.

rzezeski commented on src/riak_kv_index_hashtree.erl:
```diff
@@ -0,0 +1,592 @@
+            {reply, max_concurrency, State};
+        ok ->
+            case do_get_lock(remote_fsm, FsmPid, State) of
+                {ok, State2} ->
```
Why bother doing an explicit match against `ok` here when the 2nd match does the same thing for any reply.  I.e. nothing different is done between these two matches.

``` erlang
{Reply, State2} = do_get_lock(...),
{reply, Reply, State2}
```

rzezeski commented on src/riak_kv_index_hashtree.erl:
```diff
@@ -0,0 +1,592 @@
+                                 ok ->
+                                     build_or_rehash(Self, State)
+                             end
+                     end),
```
The building of the hashtree in `maybe_build` links the building
process to the index hashtree.  If this build crashes then the index
hashtree crashes.  Since the index hashtree is linked to the vnode it
then will also cause the vnode to crash.  Do we want this behavior?
Also, since this is a spawn-linked process it does not sit under a
supervisor.  Have you considered having a simple one-for-one that
these pids can live under?  That would make it easier to see all the
hashtree builds that are running and kill them if necessary.

rzezeski commented on src/riak_kv_index_hashtree.erl:
```diff
@@ -0,0 +1,592 @@
+fold_keys(Partition, Tree) ->
+    {ok, Ring} = riak_core_ring_manager:get_my_ring(),
+    Req = ?FOLD_REQ{foldfun=fun(BKey={Bucket,Key}, RObj, _) ->
+                                    IndexN = riak_kv_util:get_index_n({Bucket, Key}, Ring),
```
I think the `IndexN` calculation could possibly be made more efficient
by making use of the techniques found in `riak_core_repair`.  It would
avoid a `div` and `rem` operation per key which could maybe make a
difference in the tight loop of vnode folds.  Anyways, I realize the
focus right now is to make it work first, worry about optimization
later.  If I have time I might try taking a closer look at this.

jtuple committed:
Change several gen_server:calls to have infinity timeout

jtuple committed:
Remove AAE debug information stored in process dictionary

jtuple committed:
Make entropy manager handle remote locking, not remote vnode

Change the start_exchange_remote logic to route through the remote node
entropy manager rather than the remote vnode. Also, change the code to
rely upon chained casts rather than calls. Specifically, first ask the
remote entropy manager to acquire a concurrency lock. If successful, the
entropy manager than forwards the message to the relevant index hashtree
to acquire the tree lock. If successful, the index hashtree sends a reply
back to the originating exchange FSM.

Make entropy manager always run, even when AAE is disabled. This is
necessary to handle the new remote exchange process, as well as sets the
stage for runtime enable/disable of AAE.

jtuple committed:
Make AAE exchange FSM wait indefinitely for trees to update

The exchange FSM will now wait for both the local and remote hashtrees
to perform a tree update, rather than potentially timing out. Previously,
the exchange FSM would time out, while the updates would continue to run
asynchronously. However, since the exchange FSM is the process holding
both the concurrency and hashtree locks, future exchanges could be
triggered that initiate additional updates. If an update was started for
a tree already performing an update, the previous update would be killed
and you run the risk of a tree never successfully finishing an update.
Even if the update is for different trees, a potentially unbounded number
of concurrent updates could occur that are not limited by the concurrency
setting.

Also, changed the riak_kv_exchange_fsm:as_event to use spawn_link rather
than spawn to ensure call failures trigger crashes of the exchange FSM.

jtuple committed:
Change index_hashtree:do_compare to use spawn_link

jtuple committed:
Make the entropy manager not block to get hashtree PIDs

Instead of having the entropy manager use riak_kv_vnode:hashtree_pid,
which uses a synchronous vnode request, have the entropy manager use
the new riak_kv_vnode:request_hashtree_pid which sends an asynchronous
request to the vnode. The vnode will eventually reply by sending a
message back to the entropy manager which is handled by handle_info.
These changes ensure the entropy manager never blocks when there are
slow vnodes.

jtuple committed:
Ensure that index hashtree failures do not crash vnodes

Previously, riak_kv_vnode would create riak_kv_index_hashtree PIDs
using spawn_link, which would force the vnode to crash if the hashtree
ever crashed. Now, the vnode spawns and monitors the hashtree PID and
will recreate the tree if it ever dies. If creating a new tree fails,
the vnode will retry to create the tree periodically. This handles the
case where a lingering LevelDB lock has yet to be released between
when a previous tree exited and the new tree was created. The first
create may fail, but eventually the lock is released and a new tree
started.

jtuple committed:
Add build tokens to limit hashtree builds/rebuilds

jtuple committed:
Make hashtree rehashing not require a build token

After a node requests, already built hashtrees are not rebuilt but
instead rehashed. The new build token system limited both actions, but
tree rehashing is relatively cheap and something we should not block.
Otherwise, a node that restarts will take a long time before it can
initiate AAE exchanges.

jtuple committed:
Support enabling/disabling of AAE without node restart

jtuple committed:
Make AAE exchange check riak_kv service status

Exchanges will now only occur if both the local and remote nodes are
up and have the riak_kv service running. This prevents AAE from trying
to do exchanges with nodes that are offline, or performing exchanges
with a node that is up but not handling K/V requests and therefore
would not repair anything even if read-repair was triggered.

jtuple committed:
Make AAE data root configurable

jtuple committed:
Make preflist_siblings not include specified index

jtuple committed:
Rewrite determine_max_n to reuse determine_all_n

jtuple committed:
Various minor changes to AAE

-- Fix typo in riak_kv_entropy_manager:get_lock edoc.
-- Make AAE config use proplist functions for clarity of intent.
-- Rename riak_kv_entropy_manager:trigger_exchange to enqueue_exchange.
-- Change AAE config from on/off/manual to on/off, as manual already
   works as a config modifier similar to debug.
-- Fix -spec for riak_kv_entropy_manager:do_get_lock.
-- Add additional information to the "repaired X keys" info message.

rzezeski commented on src/hashtree.erl:
```diff
@@ -0,0 +1,1031 @@
+                  SegmentPath ->
+                      SegmentPath
+              end,
+    Options = [{create_if_missing, true},
```
In the future we may want to merge these with `Opts` in case we want to tweak the level options used for AAE.

rzezeski commented on src/hashtree.erl:
```diff
@@ -0,0 +1,1031 @@
+    DataDir = case proplists:get_value(segment_path, Opts) of
+                  undefined ->
+                      Root = "/tmp/anti/level",
+                      <<P:128/integer>> = crypto:md5(term_to_binary(erlang:now())),
```
I realize this code path isn't hit by AAE but should it instead default to using the `Index` as the segment path?

rzezeski commented on src/hashtree.erl:
```diff
@@ -0,0 +1,1031 @@
+    State.
+
+-spec insert(binary(), binary(), hashtree()) -> hashtree().
+insert(Key, ObjHash, State) ->
```
nitpick: The hash could be of anything, doesn't have to be an `ObjHash`.  Naming is hard.

jtuple commented:
Pretty sure I addressed all raised issues, plus several other things that came up along the way. Organized the work such that this should be easy to review commit-by-commit.

rzezeski commented:
We still have both `add_index_exchanges` and `enqueue_exchanges` which are doing the same work.  It seems we could replace the call to `add_index_excahnges` with `enqueue_exchanges`.  Not a blocker, but thought I'd throw it out there.

NOTE: Well `add_index_exchanges` checks the mode but I don't see why that couldn't be added to `enqueue_exchanges`.

rzezeski commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -0,0 +1,671 @@
+                trees          = []        :: [{index(), pid()}],
+                tree_queue     = []        :: [{index(), pid()}],
+                locks          = []        :: [{pid(), reference()}],
+                build_tokens   = 0         :: pos_integer(),
```
Dialyzer bitches about this.  Use non_neg_integer or integer.

rzezeski committed:
Fix build_tokens spec

rzezeski committed:
Fix hashtree_eqc

rzezeski commented:
When we talked in the back-channel you said this would be enabled by default.  I think it should be.  We need a Riak PR for that I guess.

jtuple committed:
Increase AAE default timeouts/limits

jtuple committed:
Enable AAE by default in riak_kv.app.src

rzezeski commented:
+1 to merge
- Reviewed code multiple times
- Compiled, ran unit tests, ran quick check
- Used new interceptor code to try emulate dropped puts and slow tree updates, AAE still worked
- Ran almost all riak tests with AAE enabled and they all passed

jtuple commented:
We'll want to do a separate pull-request to Riak to bring AAE app.config defaults into play. Take built-in timer defaults and put them into app.config as documentation, plus (obviously) enable AAE default (which is already in riak_kv.app so it'll be enabled if missing), plus anti_entropy_data_dir. Right now, we use fallback:

```
[warning] Config riak_kv/anti_entropy_data_dir is missing. Defaulting to: "./data/anti_entropy
```

All easy to change in app.config though. Might even consider adding them riak_kv.app as well, but that's up for discussion. Merging now.

DeadZen commented on docs/hashtree.md:
```diff
@@ -0,0 +1,64 @@
+key written is the binary `<<$s, SegmentId:64/integer,
+Key/binary>>`. Thus, inserting a new key-value hash is nothing more
+than a single LevelDB write. Likewise, key-hash pairs for a segment
+are laided on sequentially on-disk based on key sorting. An in-memory
```
This should read... "are laid sequentially on-disk"

DeadZen commented on docs/hashtree.md:
```diff
@@ -0,0 +1,64 @@
+
+When updating the segment hashes, a LevelDB iterator is used to access
+the segment keys in-order. The iterator seeks to the beginning of the
+segment and then iterators through all of the key-hash pairs. As an
```
This should read... "and then iterates"

DeadZen commented on docs/hashtree.md:
```diff
@@ -0,0 +1,64 @@
+optimization, the iteration process is designed to read in multiple
+segments when possible. For example, if the list of dirty segments was
+`[1, 2, 3, 5, 6, 10]`, the code will seek an iterator to the beginning
+of segment 1, iterator through all of its keys, compute the
```
This should read... "iterate through"

DeadZen commented on docs/hashtree.md:
```diff
@@ -0,0 +1,64 @@
+in a key exchange of the keys in the associated differing segments.
+
+By default, the hash tree itself is entirely in-memory. However, the
+code provides a `MEM_LEVEL` paramemter that specifics that levels
```
Typo on "parameter", this should read... "parameter that specifies that"

DeadZen commented on docs/hashtree.md:
```diff
@@ -0,0 +1,64 @@
+
+The default settings use `1024*1024` segments with a tree width of
+`1024`. Thus, the resulting tree is only 3 levels deep. And there
+are only `1+1024+1024*1024` hashs stored in memory -- so, a few
```
Typo on "hashes"
